Added logging topics for error handling

KevGoDev opened the pull request.

KevGoDev commented:
@isra17 WDYT?

This PR allows us to specify a logging topic with severity including an ignorable severity.

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,56 @@
+
+    @dataclasses.dataclass
+    class Options:
+        channel: Optional[str] = None
```
Why not use the enum? Also, I would rather use `level` instead of `channel`?

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,56 @@
+    class Options:
+        channel: Optional[str] = None
+
+    def resolve_channel_to_fn(self) -> Callable[[Any], None]:
```
you can use `logger.log(level, msg)`, then you just need to map your enum to logging level. You could actually do that straight in the enum:

```
class Level(Enum):
  IGNORE = None
  DEBUG = logging.DEBUG
  ...
```

```
if self.options.level is None:
  return
self.logger.log(self.options.level, ...)
```

KevGoDev replied:
Ok so I've changed this a bit. The `level` attibute inside `LoggingTopic.Options` is a string in order to be abke to parse the level from our yaml definitions. It then uses a mapping and stores the actual level inside of `LoggingTopic`

isra17 replied:
I'm pretty sure you can serialize enum from the options:

```python
class Level(enum.Enum):
  INFO = "INFO"
class Options:
   level: Level
```

```yaml
options:
  level: INFO
```

Did you try?

KevGoDev replied:
Ok so now it uses the enum entirely. When passing the value to `logging.Logger.log` I use the function `logging.getLevelName(self.level.value)` to resolve it to an int, which is required by `logging.Logger.log`.

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,56 @@
+
+    def __init__(self, options: Options, **kwargs: object) -> None:
+        self.channel = options.channel or self.Channels.IGNORE.value
+        self.logger = getLogger(__name__, self)
```
use the topic name in the logger name? (And don't use our own `getLogger` helper)

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,56 @@
+        yield
+
+    async def publish(self, message: TopicMessage, wait: bool) -> bool:
+        self.logger_fn(json.dumps({"id": message.id, "args": message.args}))
```
structured data must go to `extra` argument (Check elsewhere in the code). Keep message human readable.
Since the main use case of this is to dump exception, I would do the extra work to check if the message is an error message and format with a backtrace and all (And pass it to extra through `exception`, check the logger service, there's a way exception are extracted from sys.exc_info and pass to extra.

KevGoDev replied:
I can't seem to find any reference to pass "exception" inside extra. 

We could however do the following:

```python
async def publish(self, message: TopicMessage, wait: bool) -> bool:
    if self.options.level is None:
        return True
    
    extra = {"data":{"id": message.id, "args": message.args}}
    
    exc_info = None
    status = "success"
    if "traceback" in message.args:
        status = "error"
        exc_info = sys.exc_info()

    self.logger.log(
        self.options.level.value,
        f"Message executed with {status}",
        extra=extra,
        exc_info=exc_info,
    )
    return True
```

wdyt?

isra17 replied:
Right, it seems to be logger specific. In that case I would dump the backtrace of the error into `extra={"data": {"exception": backtrace}}`, I think it should give something similar to what `.exception` would give.

isra17 replied:
And as the message, `f{message.args["type"]}: f{message.args["message"]}`

isra17 replied:
> Right, it seems to be logger specific. In that case I would dump the backtrace of the error into `extra={"data": {"exception": backtrace}}`, I think it should give something similar to what `.exception` would give.

Just to clear, I mean: `backtrace = message.args["traceback"].format()`. Also make sure to only do this when it's an error message! Otherwise, might just want to log the message id and pass the message args in extra.data.

isra17 reviewed (changes requested):
I like the concept, but I would do a few change to how we log. There's a way we can pass exception backtrace in `extra` and have a similar behavior than the default logger for unhandled exception.

That's pretty much what I would aim for: Same behavior as default logger, except we can override the logger name and level.

KevGoDev commented:
@isra17 Updated PR. Added 2 examples that use the error logging topic and an ignorable logging topic (dev/null).

I also changed a bit how we pass the errors using the TopicMessage. Instead of using `dataclassses.todict` to match `TopicMessage.args` I simply pass `args={"error": ErrorMessageArgs(...)}`.
This allows us to keep our more complex types like `TracebackData` while maintaining serializability. 

Let me know what you think!

isra17 commented on src/saturn_engine/utils/tester/json_utils.py:
```diff
@@ -1,5 +1,18 @@
+import dataclasses
+
+
+def normalize_json(obj: object) -> object:
```
nit, but you can use [`@overload`](https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading) to make it so it returns a list if it take in a list (And a dict if it takes in a dict). This way you won't have to `cast` thing later on.

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,71 @@
+        if self.level == Level.IGNORE:
+            return True
+
+        extra = {"data": {"id": message.id, "args": message.args}}
```
import and use `topic_message_data` from the logging service ?

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,71 @@
+            return True
+
+        extra = {"data": {"id": message.id, "args": message.args}}
+        msg = f"Message {message.id} executed successfully"
```
This is `publish`, `message` wasn't executed, it's being published.

nit, good practice is to isolate the varying part of the message to make it easier to write log query/regex, aka:

```
f"Message published: {message.id}"
```

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,71 @@
+            # remove args from extra since it would be redundant with the traceback
+            del extra["data"]["args"]
+            # get our error message and put the traceback in our extra data
+            error_msg: ErrorMessageArgs = cast(ErrorMessageArgs, message.args["error"])
```
don't cast, `isinstance` instead. You might want to do the `isinstance` before the `del`, to make sure we really are dealing with an `ErrorMessageArgs` and not some other `error` arg. aka

```python
if isinstance(message.args.get("error"), ErrorMessageArgs):
  # do stuff here
```

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,71 @@
+            error_msg: ErrorMessageArgs = cast(ErrorMessageArgs, message.args["error"])
+            msg = error_msg.message
+            extra["data"]["exception"] = "".join(
+                [line for line in error_msg.traceback.format()]
```
just `"".join(error_msg.traceback.format())`

isra17 replied:
I would actually suggest to: `str(RemoteException(error_msg.traceback))`. It has the standard exception formatting.

isra17 commented on tests/worker/topics/test_logging_topic.py:
```diff
@@ -0,0 +1,134 @@
+        },
+    )
+
+    log_patch = mocker.patch("logging.Logger.log", return_value=None)
```
Don't mock, `pytest` has `caplog` for that (See `"tests/worker/services/test_logger.py"`)

KevGoDev commented:
@isra17 Updated PR with requested changes!

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,61 @@
+            extra["data"]["exception"] = str(RemoteException(error_msg.traceback))
+
+        self.logger.log(
+            logging.getLevelName(self.level.value),
```
That should be made in the `__init__` (`self.level = logging.getLevelName`)

isra17 replied:
This is going to make `IGNORE` unhandlable, but I would probably just drop `IGNORE`. If we want to ignore, we should use a `NullTopic` or something like that.

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,61 @@
+        error_msg = message.args.get("error")
+        if isinstance(error_msg, ErrorMessageArgs):
+            # remove args from extra since it would be redundant with the traceback
+            del extra["data"]["args"]
```
nit, maybe just pop the error?

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,61 @@
+        self.logger = logging.getLogger(f"saturn.topics.{self.name}")
+
+    async def run(self) -> AsyncGenerator[TopicMessage, None]:
+        return
```
`raise NotImplementedError` would be more appropriate?

isra17 reviewed (changes requested):
One last review round and I believe this is going to be merged afterward :)

KevGoDev commented:
@isra17 Removed the `IGNORE` level for the logging topic and added a `NullTopic` instead.

isra17 commented on src/saturn_engine/worker/topics/logger.py:
```diff
@@ -0,0 +1,57 @@
+        error_msg = message.args.get("error")
+        if isinstance(error_msg, ErrorMessageArgs):
+            # remove args from extra since it would be redundant with the traceback
+            extra["data"].pop("args")
```
nit, but I would even go as far as `extra["data"]["args"].pop("error")`

isra17 reviewed (approved):
lgtm!
